File: javascript/rabbit_feeder.js
#!/usr/bin/env node

var r = require('rethinkdb');
var amqp = require('amqplib');

var rethinkConn;
var rabbitConn;
var channel;
var exchange = 'rethinkdb';


r.connect({host: 'localhost', port: 28015}).then(function(conn) {
    // Setup RethinkDB connection
    rethinkConn = conn;
}).catch(r.Error.RqlDriverError, function(err){
    console.log(err.message);
    process.exit(1);
}).then(function createDB(){
    // Ensure database exists for this example
    return r.dbCreate('change_example').run(rethinkConn);
}).finally(function createTable(){
    // Ensure table exists for this example
    return r.db('change_example').tableCreate('mytable').run(rethinkConn);
}).catch(r.Error.RqlRuntimeError, function(){
    // We ignore db/table exists errors here because amqplib uses a
    // different promise library from rethinkdb, and the error doesn't
    // propagate correctly from one to the other.
}).then(function(){
    // Setup rabbit connection
    return amqp.connect('amqp://localhost:5672');
}).then(function(conn){
    rabbitConn = conn;
    // Setup rabbit channel
    return rabbitConn.createChannel();
}).then(function(ch){
    channel = ch;
    // Setup rabbit exchange
    return channel.assertExchange(exchange, 'topic', {durable: false});
}).then(function(){
    // Listen for changes on our table
    return r.db('change_example').table('mytable').changes().run(rethinkConn);
}).then(function(changeCursor){
    // Feed changes into rabbit
    changeCursor.each(function(err, change){
        if(err){
            // The table may have been deleted, or possibly connection issues
            console.log('A problem reading from RethinkDB cursor:');
            console.log(err.msg);
            process.exit(1);
        }
        var routingKey = 'mytable.' + typeOfChange(change);
        console.log('RethinkDB -(', routingKey, ')-> RabbitMQ')
        channel.publish(
            exchange, routingKey, new Buffer(JSON.stringify(change)));
    });
}).catch(function(err){
    console.log(err.message);
    process.exit(1);
});

function typeOfChange(change) {
    // Determines whether the change is a create, delete or update
    if(change.old_val === null){
        return 'create';
    } else if(change.new_val === null){
        return 'delete';
    } else {
        return 'update';
    }
    return 'something'
}

File: ruby/rabbit_feeder.rb
#!/usr/bin/env ruby

# Listens for changes on a rethinkdb table and sends them over a
# RabbitMQ topic exchange.
#
# The topic will be <tablename>.<type_of_change> where
# type_of_change is create, delete or update.

require 'bunny'
require 'rethinkdb'
include RethinkDB::Shortcuts
require 'json'

# Setup RethinkDB connection
rethink_conn = r.connect(:host => 'localhost', :port => 28015)

# Ensure correct database and table exist for this example
begin
  r.db_create("change_example").run(rethink_conn)
rescue RethinkDB::RqlRuntimeError
end

begin
  r.db("change_example").table_create("mytable").run(rethink_conn)
rescue RethinkDB::RqlRuntimeError
end

# Setup rabbit connection and exchange
rabbit_conn = Bunny.new(:host => 'localhost', :port => 5672).start
channel = rabbit_conn.create_channel
exchange = channel.topic("rethinkdb")

# Determines whether the change is a create, delete or update
def type_of_change(change)
  if change['old_val'].nil?
    :create
  elsif change['new_val'].nil?
    :delete
  else
    :update
  end
end

# Start feeding...
table_changes = r.db('change_example').table('mytable').changes()

begin
  table_changes.run(rethink_conn).each do |change|
    routing_key = "mytable.#{type_of_change change}"
    puts "RethinkDB -( #{routing_key} )-> RabbitMQ"
    exchange.publish(change.to_json, :routing_key => routing_key)
  end
rescue RethinkDB::RqlRuntimeError => e
  # Table may have been dropped, connection failed etc
  puts e.message
end
